File: engram-ts/src/memory.ts
/**
 * Engram Memory — Public API
 */

import { MemoryConfig } from './config';
import { MemoryEntry, MemoryType, MemoryLayer, DEFAULT_IMPORTANCE } from './core';
import { SQLiteStore } from './store';
import { retrieveTopK } from './activation';
import { SearchEngine, SearchResult } from './search';
import { runConsolidationCycle, getConsolidationStats } from './consolidation';
import { effectiveStrength, shouldForget, pruneforgotten } from './forgetting';
import { confidenceScore, confidenceLabel } from './confidence';
import { detectFeedback, applyReward } from './reward';
import { synapticDownscale } from './downscaling';
import { BaselineTracker } from './anomaly';
import { recordCoactivation, decayHebbianLinks, getHebbianNeighbors } from './hebbian';
import { SessionWorkingMemory, SessionRecallResult, getSessionWM } from './session_wm';
import { EmbeddingProvider, EmbeddingConfig, DEFAULT_EMBEDDING_CONFIG } from './embeddings/base';
import { detectProvider, getAvailableProviders } from './embeddings/provider_detection';
import { migrateVectorColumn, storeVector, getVectorCount } from './vector_search';
import { hybridSearch, adaptiveHybridSearch } from './hybrid_search';

const TYPE_MAP: Record<string, MemoryType> = {};
for (const t of Object.values(MemoryType)) {
  TYPE_MAP[t] = t;
}

export class Memory {
  path: string;
  config: MemoryConfig;
  _store: SQLiteStore;
  private _tracker: BaselineTracker;
  private _createdAt: number;
  private _embeddingProvider: EmbeddingProvider | null = null;
  private _embeddingConfig: EmbeddingConfig;
  private _embeddingInitialized = false;

  constructor(path: string = './engram.db', config?: MemoryConfig, embeddingConfig?: EmbeddingConfig) {
    this.path = path;
    this.config = config ?? MemoryConfig.default();
    this._store = new SQLiteStore(path);
    this._tracker = new BaselineTracker(this.config.anomalyWindowSize);
    this._createdAt = Date.now() / 1000;
    this._embeddingConfig = embeddingConfig ?? DEFAULT_EMBEDDING_CONFIG;
    
    // Migrate vector column if needed
    migrateVectorColumn(this._store.db);
  }

  /**
   * Lazily initialize embedding provider
   */
  private async _ensureEmbedding(): Promise<void> {
    if (this._embeddingInitialized) return;
    
    this._embeddingProvider = await detectProvider(this._embeddingConfig);
    this._embeddingInitialized = true;
  }

  /**
   * Add a new memory (synchronous, no embedding)
   * For embedding support, use `addWithEmbedding()` instead
   */
  add(
    content: string,
    opts: {
      type?: string;
      importance?: number;
      source?: string;
      tags?: string[];
      entities?: Array<string | [string, string]>;
      contradicts?: string;
    } = {},
  ): string {
    const {
      type = 'factual',
      importance,
      source = '',
      tags,
      entities,
      contradicts,
    } = opts;

    const memoryType = TYPE_MAP[type] ?? MemoryType.FACTUAL;

    let actualContent = content;
    if (tags && tags.length > 0) {
      actualContent = `${content} [tags: ${tags.join(', ')}]`;
    }

    const entry = this._store.add(actualContent, memoryType, importance, source);

    if (contradicts) {
      const oldEntry = this._store.get(contradicts);
      if (oldEntry) {
        entry.contradicts = contradicts;
        this._store.update(entry);
        oldEntry.contradictedBy = entry.id;
        this._store.update(oldEntry);
      }
    }

    if (entities) {
      for (const ent of entities) {
        if (Array.isArray(ent)) {
          const [entity, relation] = ent;
          this._store.addGraphLink(entry.id, entity, relation ?? '');
        } else {
          this._store.addGraphLink(entry.id, ent, '');
        }
      }
    }

    this._tracker.update('encoding_rate', 1.0);
    return entry.id;
  }

  /**
   * Add a new memory with embedding support (async)
   */
  async addWithEmbedding(
    content: string,
    opts: {
      type?: string;
      importance?: number;
      source?: string;
      tags?: string[];
      entities?: Array<string | [string, string]>;
      contradicts?: string;
    } = {},
  ): Promise<string> {
    // Add memory first (synchronous)
    const memoryId = this.add(content, opts);

    // Generate and store embedding (async)
    await this._ensureEmbedding();
    
    if (this._embeddingProvider && this._embeddingProvider.name !== 'none') {
      try {
        const result = await this._embeddingProvider.embed(content);
        storeVector(this._store.db, memoryId, result.embedding);
      } catch (error) {
        console.error(`Failed to generate embedding for ${memoryId}:`, error);
        // Continue without embedding (graceful degradation)
      }
    }

    return memoryId;
  }

  recall(
    query: string,
    opts: {
      limit?: number;
      context?: string[];
      types?: string[];
      minConfidence?: number;
      graphExpand?: boolean;
    } = {},
  ): Array<{
    id: string;
    content: string;
    type: string;
    confidence: number;
    confidence_label: string;
    strength: number;
    activation: number;
    age_days: number;
    layer: string;
    importance: number;
    contradicted: boolean;
  }> {
    const {
      limit = 5,
      context,
      types,
      minConfidence = 0.0,
      graphExpand = true,
    } = opts;

    const engine = new SearchEngine(this._store);
    const searchResults = engine.search({
      query,
      limit,
      contextKeywords: context,
      types,
      minConfidence,
      graphExpand,
    });

    const now = Date.now() / 1000;
    const output = searchResults.map(r => ({
      id: r.entry.id,
      content: r.entry.content,
      type: r.entry.memoryType,
      confidence: Math.round(r.confidence * 1000) / 1000,
      confidence_label: r.confidenceLabel,
      strength: Math.round(effectiveStrength(r.entry, now) * 1000) / 1000,
      activation: Math.round(r.score * 1000) / 1000,
      age_days: Math.round(r.entry.ageDays() * 10) / 10,
      layer: r.entry.layer,
      importance: Math.round(r.entry.importance * 100) / 100,
      contradicted: Boolean(r.entry.contradictedBy),
    }));

    // Record Hebbian co-activation
    if (this.config.hebbianEnabled && searchResults.length >= 2) {
      const resultIds = searchResults.map(r => r.entry.id);
      recordCoactivation(this._store, resultIds, this.config);
    }

    this._tracker.update('retrieval_count', output.length);
    return output;
  }

  /**
   * Recall with embedding support (hybrid search)
   * Combines vector similarity + FTS5 for better cross-language/semantic recall
   */
  async recallWithEmbedding(
    query: string,
    opts: {
      limit?: number;
      context?: string[];
      types?: string[];
      minConfidence?: number;
      vectorWeight?: number;
      ftsWeight?: number;
    } = {},
  ): Promise<Array<{
    id: string;
    content: string;
    type: string;
    confidence: number;
    confidence_label: string;
    strength: number;
    activation: number;
    age_days: number;
    layer: string;
    importance: number;
    contradicted: boolean;
    vector_score?: number;
    fts_score?: number;
  }>> {
    const {
      limit = 5,
      vectorWeight = 0.7,
      ftsWeight = 0.3,
    } = opts;

    // Ensure embedding provider initialized
    await this._ensureEmbedding();

    // Generate query embedding
    let queryVector: number[] | null = null;
    if (this._embeddingProvider && this._embeddingProvider.name !== 'none') {
      try {
        const result = await this._embeddingProvider.embed(query);
        queryVector = result.embedding;
      } catch (error) {
        console.error('Failed to generate query embedding:', error);
        // Fall back to FTS5 only
      }
    }

    // Hybrid search
    const searchResults = adaptiveHybridSearch(
      this._store.db,
      queryVector,
      query,
      limit,
    );

    // Convert to Memory format
    const now = Date.now() / 1000;
    const output = searchResults.map(r => {
      const entry = this._store.get(r.id);
      if (!entry) {
        throw new Error(`Memory ${r.id} not found`);
      }

      const conf = confidenceScore(entry, null, now);

      return {
        id: entry.id,
        content: entry.content,
        type: entry.memoryType,
        confidence: Math.round(conf * 1000) / 1000,
        confidence_label: confidenceLabel(conf),
        strength: Math.round(effectiveStrength(entry, now) * 1000) / 1000,
        activation: Math.round(r.score * 1000) / 1000,
        age_days: Math.round(entry.ageDays() * 10) / 10,
        layer: entry.layer,
        importance: Math.round(entry.importance * 100) / 100,
        contradicted: Boolean(entry.contradictedBy),
        vector_score: Math.round(r.vectorScore * 1000) / 1000,
        fts_score: Math.round(r.ftsScore * 1000) / 1000,
      };
    });

    // Record Hebbian co-activation
    if (this.config.hebbianEnabled && output.length >= 2) {
      const resultIds = output.map(r => r.id);
      recordCoactivation(this._store, resultIds, this.config);
    }

    this._tracker.update('retrieval_count', output.length);
    return output;
  }

  consolidate(days: number = 1.0): void {
    runConsolidationCycle(
      this._store,
      days,
      this.config.interleaveRatio,
      this.config.alpha,
      this.config.mu1,
      this.config.mu2,
      this.config.replayBoost,
      this.config.promoteThreshold,
      this.config.demoteThreshold,
      this.config.archiveThreshold,
    );
    synapticDownscale(this._store, this.config.downscaleFactor);
    
    // Decay Hebbian links
    if (this.config.hebbianEnabled) {
      decayHebbianLinks(this._store, this.config.hebbianDecay);
    }
  }

  forget(opts: { memoryId?: string; threshold?: number } = {}): void {
    const threshold = opts.threshold ?? this.config.forgetThreshold;
    if (opts.memoryId) {
      this._store.delete(opts.memoryId);
    } else {
      pruneforgotten(this._store, threshold);
    }
  }

  reward(feedback: string, recentN: number = 3): void {
    const [polarity, conf] = detectFeedback(feedback);
    if (polarity === 'neutral' || conf < 0.3) return;
    applyReward(this._store, polarity, recentN, this.config.rewardMagnitude * conf);
  }

  downscale(factor?: number): { n_scaled: number; avg_before: number; avg_after: number } {
    return synapticDownscale(this._store, factor ?? this.config.downscaleFactor);
  }

  stats(): Record<string, any> {
    const consolidation = getConsolidationStats(this._store);
    const allMem = this._store.all();
    const now = Date.now() / 1000;

    const byType: Record<string, { count: number; avg_strength: number; avg_importance: number }> = {};
    for (const mt of Object.values(MemoryType)) {
      const entries = allMem.filter(m => m.memoryType === mt);
      if (entries.length > 0) {
        byType[mt] = {
          count: entries.length,
          avg_strength: Math.round(
            (entries.reduce((s, m) => s + effectiveStrength(m, now), 0) / entries.length) * 1000
          ) / 1000,
          avg_importance: Math.round(
            (entries.reduce((s, m) => s + m.importance, 0) / entries.length) * 100
          ) / 100,
        };
      }
    }

    return {
      total_memories: allMem.length,
      by_type: byType,
      layers: consolidation.layers,
      pinned: consolidation.pinned,
      uptime_hours: Math.round(((now - this._createdAt) / 3600) * 10) / 10,
      anomaly_metrics: this._tracker.metrics(),
    };
  }

  export(path: string): void {
    this._store.export(path);
  }

  updateMemory(memoryId: string, newContent: string, reason: string = 'correction'): string {
    const oldEntry = this._store.get(memoryId);
    if (!oldEntry) throw new Error(`Memory ${memoryId} not found`);

    return this.add(newContent, {
      type: oldEntry.memoryType,
      importance: oldEntry.importance,
      source: `${reason}:${memoryId}`,
      contradicts: memoryId,
    });
  }

  pin(memoryId: string): void {
    const entry = this._store.get(memoryId);
    if (entry) {
      entry.pinned = true;
      this._store.update(entry);
    }
  }

  unpin(memoryId: string): void {
    const entry = this._store.get(memoryId);
    if (entry) {
      entry.pinned = false;
      this._store.update(entry);
    }
  }

  hebbianLinks(memoryId: string): string[] {
    return getHebbianNeighbors(this._store, memoryId);
  }

  /**
   * Session-aware recall using cognitive working memory model.
   *
   * Instead of always doing expensive retrieval, this:
   * 1. Checks if the query topic overlaps with current working memory
   * 2. If yes (continuous topic) → returns cached working memory items
   * 3. If no (topic switch) → does full recall and updates working memory
   *
   * Based on Miller's Law (7±2 chunks) and Baddeley's Working Memory Model.
   * Reduces API calls by 70-80% for continuous conversation topics.
   */
  sessionRecall(
    query: string,
    opts: {
      sessionId?: string;
      sessionWM?: SessionWorkingMemory;
      limit?: number;
      types?: string[];
      minConfidence?: number;
    } = {},
  ): SessionRecallResult {
    const {
      sessionId = 'default',
      sessionWM,
      limit = 5,
      types,
      minConfidence = 0.0,
    } = opts;

    const swm = sessionWM ?? getSessionWM(sessionId);
    const wasEmpty = swm.isEmpty();
    const needsFull = wasEmpty || swm.needsRecall(query, this);

    let results: Array<{
      id: string;
      content: string;
      type: string;
      confidence: number;
      confidence_label: string;
      strength: number;
      age_days: number;
      from_working_memory: boolean;
    }>;

    if (needsFull) {
      // Full recall
      const recallResults = this.recall(query, { limit, types, minConfidence });
      results = recallResults.map(r => ({
        id: r.id,
        content: r.content,
        type: r.type,
        confidence: r.confidence,
        confidence_label: r.confidence_label,
        strength: r.strength,
        age_days: r.age_days,
        from_working_memory: false,
      }));

      // Update working memory
      swm.activate(results.map(r => r.id));
    } else {
      // Return working memory items
      const wmItems = swm.getActiveMemories(this);
      results = wmItems.map(r => ({
        id: r.id,
        content: r.content,
        type: r.type,
        confidence: r.confidence,
        confidence_label: r.confidence_label,
        strength: r.strength,
        age_days: r.age_days,
        from_working_memory: true,
      }));
    }

    return {
      results,
      fullRecallTriggered: needsFull,
      workingMemorySize: swm.size(),
      reason: wasEmpty ? 'empty_wm' : (needsFull ? 'topic_change' : 'topic_continuous'),
    };
  }

  /**
   * Get embedding provider status
   */
  async embeddingStatus(): Promise<{
    provider: string;
    model: string;
    dimensions: number;
    available: boolean;
    vector_count: number;
    available_providers: {
      ollama: boolean;
      mcp: boolean;
      openai: boolean;
      selected: string;
    };
    error?: string;
  }> {
    await this._ensureEmbedding();

    const vectorCount = getVectorCount(this._store.db);
    const availableProviders = await getAvailableProviders(this._embeddingConfig);

    if (!this._embeddingProvider) {
      return {
        provider: 'none',
        model: 'none',
        dimensions: 0,
        available: false,
        vector_count: vectorCount,
        available_providers: availableProviders,
        error: 'No embedding provider configured',
      };
    }

    const info = await this._embeddingProvider.getInfo?.() || {
      name: this._embeddingProvider.name,
      model: this._embeddingProvider.model,
      dimensions: 0,
      available: false,
    };

    return {
      provider: info.name,
      model: info.model,
      dimensions: info.dimensions,
      available: info.available,
      vector_count: vectorCount,
      available_providers: availableProviders,
      error: info.error,
    };
  }

  close(): void {
    if (this._embeddingProvider && 'close' in this._embeddingProvider) {
      (this._embeddingProvider as any).close();
    }
    this._store.close();
  }

  get length(): number {
    return this._store.all().length;
  }

  toString(): string {
    return `Memory(path='${this.path}', entries=${this.length})`;
  }
}

File: engram/memory.py
"""
Engram Memory — Public API

The unified interface to the neuroscience-grounded memory system.
Designed for simplicity (like Mem0) while backed by mathematical models
from cognitive science.

Architecture:
    Memory (this class)
    ├── MemoryStore (backend — swappable to SQLiteStore)
    ├── activation.py (ACT-R retrieval)
    ├── consolidation.py (Memory Chain Model)
    ├── forgetting.py (Ebbinghaus + interference)
    ├── confidence.py (metacognitive scoring)
    ├── reward.py (dopaminergic feedback)
    ├── downscaling.py (homeostatic plasticity)
    └── anomaly.py (predictive coding)

Usage:
    from engram import Memory

    mem = Memory("./agent.db")
    mem.add("potato prefers action over discussion", type="relational", importance=0.6)
    mem.add("Use www.moltbook.com not moltbook.com", type="procedural", importance=0.8)

    results = mem.recall("what does potato prefer?", limit=5)
    for r in results:
        print(f"[{r['confidence_label']}] {r['content']}")

    mem.consolidate()  # Run "sleep" cycle
    mem.reward("good job!")  # Positive feedback strengthens recent memories
    mem.stats()
"""

import sys
import os
import time


from typing import Optional, Union

from engram.config import MemoryConfig
from engram.core import MemoryEntry, MemoryStore, MemoryType, MemoryLayer, DEFAULT_IMPORTANCE
from engram.store import SQLiteStore
from engram.activation import retrieve_top_k
from engram.search import SearchEngine
from engram.hybrid_search import HybridSearchEngine
from engram.consolidation import run_consolidation_cycle, get_consolidation_stats
from engram.forgetting import effective_strength, should_forget, prune_forgotten
from engram.confidence import confidence_score, confidence_label
from engram.reward import detect_feedback, apply_reward
from engram.downscaling import synaptic_downscale
from engram.anomaly import BaselineTracker
from engram.hebbian import (
    record_coactivation,
    get_hebbian_neighbors,
    get_all_hebbian_links,
    decay_hebbian_links,
)


# Map string type names to MemoryType enum
_TYPE_MAP = {t.value: t for t in MemoryType}


class Memory:
    """
    Main interface to the Engram memory system.

    Wraps the neuroscience math models behind a clean API.
    All complexity is hidden — you just add, recall, and consolidate.

    Backend: SQLiteStore for persistent storage with FTS5 search.
    """

    def __init__(
        self, 
        path: str = "./engram.db", 
        config: MemoryConfig = None,
        embedding = None,
    ):
        """
        Initialize Engram memory system.

        Args:
            path: Path to SQLite database file. Created if it doesn't exist.
                  Use ":memory:" for in-memory (non-persistent) operation.
            config: MemoryConfig with tunable parameters. None = literature defaults.
            embedding: Optional embedding adapter for semantic search.
                      Can be an EmbeddingAdapter instance or a string shortcut:
                      - "openai" -> OpenAIAdapter (requires OPENAI_API_KEY)
                      - "ollama" -> OllamaAdapter (requires local Ollama)
                      - None -> FTS5-only mode (no embeddings)
        """
        self.path = path
        self.config = config or MemoryConfig.default()
        self._store = SQLiteStore(path)
        self._tracker = BaselineTracker(window_size=self.config.anomaly_window_size)
        self._created_at = time.time()
        
        # Initialize embedding support
        self._embedding_adapter = None
        self._vector_store = None
        
        if embedding is not None:
            self._init_embedding(embedding)

    def _init_embedding(self, embedding):
        """Initialize embedding adapter and vector store."""
        from engram.vector_store import VectorStore
        
        # Handle string shortcuts
        if isinstance(embedding, str):
            if embedding == "openai":
                from engram.embeddings import OpenAIAdapter
                self._embedding_adapter = OpenAIAdapter()
            elif embedding == "ollama":
                from engram.embeddings import OllamaAdapter
                self._embedding_adapter = OllamaAdapter()
            else:
                raise ValueError(f"Unknown embedding shortcut: {embedding}. Use 'openai', 'ollama', or pass an adapter instance.")
        else:
            # Assume it's an adapter instance
            self._embedding_adapter = embedding
        
        # Initialize vector store
        self._vector_store = VectorStore(self._store._conn, self._embedding_adapter)

    def add(self, content: str, type: str = "factual", importance: float = None,
            source: str = "", tags: list[str] = None,
            entities: list = None, contradicts: str = None,
            created_at: float = None) -> str:
        """
        Store a new memory. Returns memory ID.

        The memory is encoded with initial working_strength=1.0 (strong
        hippocampal trace) and core_strength=0.0 (no neocortical trace yet).
        Consolidation cycles will gradually transfer it to core.

        Importance modulates encoding strength — high importance memories
        (emotional, relational) are encoded more strongly, mimicking the
        amygdala's role in memory modulation.

        Args:
            content: The memory content (natural language)
            type: Memory type — one of: factual, episodic, relational,
                  emotional, procedural, opinion
            importance: 0-1 importance score (None = auto from type)
            source: Source identifier (e.g., filename, conversation ID)
            tags: Optional tags for categorization (stored in content for now)

        Returns:
            Memory ID string (8-char UUID prefix)
        """
        memory_type = _TYPE_MAP.get(type, MemoryType.FACTUAL)

        # If tags provided, append to content for searchability
        actual_content = content
        if tags:
            actual_content = f"{content} [tags: {', '.join(tags)}]"

        entry = self._store.add(
            content=actual_content,
            memory_type=memory_type,
            importance=importance,
            source_file=source,
            created_at=created_at,
        )

        # Handle contradiction linking
        if contradicts:
            old_entry = self._store.get(contradicts)
            if old_entry:
                entry.contradicts = contradicts
                # Update entry in store with contradiction field
                self._store.update(entry)
                # Mark old memory as contradicted
                old_entry.contradicted_by = entry.id
                self._store.update(old_entry)

        # Store graph links if entities provided
        if entities:
            for ent in entities:
                if isinstance(ent, (list, tuple)):
                    entity, relation = ent[0], ent[1] if len(ent) > 1 else ""
                else:
                    entity, relation = ent, ""
                self._store.add_graph_link(entry.id, entity, relation)

        # Track encoding rate for anomaly detection
        self._tracker.update("encoding_rate", 1.0)
        
        # Store embedding if adapter is configured
        if self._vector_store is not None:
            self._vector_store.add(entry.id, actual_content)

        return entry.id

    def recall(self, query: str, limit: int = 5,
               context: list[str] = None,
               types: list[str] = None,
               min_confidence: float = 0.0,
               graph_expand: bool = True) -> list[dict]:
        """
        Retrieve relevant memories using ACT-R activation-based retrieval.

        Unlike simple cosine similarity, this uses:
        - Base-level activation (frequency × recency, power law)
        - Spreading activation from context keywords
        - Importance modulation (emotional memories are more accessible)

        Results include a confidence score (metacognitive monitoring)
        that tells you how "trustworthy" each retrieval is.

        Args:
            query: Natural language query
            limit: Maximum number of results
            context: Additional context keywords to boost relevant memories
            types: Filter by memory types (e.g., ["factual", "procedural"])
            min_confidence: Minimum confidence threshold (0-1)

        Returns:
            List of dicts: {id, content, type, confidence, confidence_label,
                           strength, age_days, layer, importance}
        """
        # Use hybrid search if embeddings are available, else FTS5-only
        if self._vector_store is not None:
            engine = HybridSearchEngine(self._store, self._vector_store)
        else:
            engine = SearchEngine(self._store)
        
        search_results = engine.search(
            query=query,
            limit=limit,
            context_keywords=context,
            types=types,
            min_confidence=min_confidence,
            graph_expand=graph_expand,
        )

        output = []
        for r in search_results:
            output.append({
                "id": r.entry.id,
                "content": r.entry.content,
                "type": r.entry.memory_type.value,
                "confidence": round(r.confidence, 3),
                "confidence_label": r.confidence_label,
                "strength": round(effective_strength(r.entry, now=time.time()), 3),
                "activation": round(r.score, 3),
                "age_days": round(r.entry.age_days(), 1),
                "layer": r.entry.layer.value,
                "importance": round(r.entry.importance, 2),
                "contradicted": bool(r.entry.contradicted_by),
                "source": r.entry.source_file,  # Include source for evidence matching
            })

        # Track retrieval for anomaly detection
        self._tracker.update("retrieval_count", len(output))

        # ACT-R: Record access for all retrieved memories (boosts future retrieval)
        for r in search_results:
            self._store.record_access(r.entry.id)

        # Hebbian learning: record co-activation for recalled memories
        if self.config.hebbian_enabled and len(output) >= 2:
            memory_ids = [r["id"] for r in output]
            record_coactivation(
                self._store,
                memory_ids,
                threshold=self.config.hebbian_threshold,
            )

        return output

    def consolidate(self, days: float = 1.0):
        """
        Run a consolidation cycle ("sleep replay").

        This is the core of memory maintenance. Based on Murre & Chessa's
        Memory Chain Model, it:

        1. Decays working_strength (hippocampal traces fade)
        2. Transfers knowledge to core_strength (neocortical consolidation)
        3. Replays archived memories (prevents catastrophic forgetting)
        4. Rebalances layers (promote strong → core, demote weak → archive)

        Call this periodically — once per "day" of agent operation,
        or after significant learning sessions.

        Also runs synaptic downscaling to prevent unbounded strength growth.

        Args:
            days: Simulated time step in days (1.0 = one day of consolidation)
        """
        run_consolidation_cycle(
            self._store, dt_days=days,
            interleave_ratio=self.config.interleave_ratio,
            alpha=self.config.alpha,
            mu1=self.config.mu1, mu2=self.config.mu2,
            replay_boost=self.config.replay_boost,
            promote_threshold=self.config.promote_threshold,
            demote_threshold=self.config.demote_threshold,
            archive_threshold=self.config.archive_threshold,
        )
        synaptic_downscale(self._store, factor=self.config.downscale_factor)

        # Decay Hebbian links during consolidation
        if self.config.hebbian_enabled:
            decay_hebbian_links(self._store, factor=self.config.hebbian_decay)

    def forget(self, memory_id: str = None, threshold: float = None):
        """
        Forget a specific memory or prune all below threshold.

        If memory_id is given, removes that specific memory.
        Otherwise, prunes all memories whose effective_strength
        is below threshold (moves them to archive).

        This mirrors natural forgetting — memories aren't truly deleted,
        they become inaccessible (archived). They could theoretically
        be recovered with the right retrieval cue.

        Args:
            memory_id: Specific memory to forget (None = prune all weak)
            threshold: Strength threshold for pruning (default 0.01)
        """
        if threshold is None:
            threshold = self.config.forget_threshold
        if memory_id is not None:
            self._store.delete(memory_id)
        else:
            prune_forgotten(self._store, threshold=threshold)

    def reward(self, feedback: str, recent_n: int = 3):
        """
        Process user feedback as a dopaminergic reward signal.

        Detects positive/negative sentiment in the feedback text,
        then applies reward modulation to the N most recently
        accessed memories. This shapes future behavior:
        - Positive → memories consolidate faster, more retrievable
        - Negative → memories suppressed, less likely to influence output

        Args:
            feedback: Natural language feedback from user
            recent_n: Number of recent memories to affect
        """
        polarity, conf = detect_feedback(feedback)

        if polarity == "neutral" or conf < 0.3:
            return  # Not confident enough to act

        apply_reward(self._store, polarity, recent_n=recent_n,
                     reward_magnitude=self.config.reward_magnitude * conf)

    def downscale(self, factor: float = None):
        """
        Global synaptic downscaling — normalize all memory weights.

        Based on Tononi & Cirelli's Synaptic Homeostasis Hypothesis:
        during sleep, all synaptic weights are proportionally reduced.
        This prevents unbounded growth and maintains discriminability.

        Args:
            factor: Multiplicative factor (0-1). Default 0.95 = 5% reduction.

        Returns:
            Stats dict: {n_scaled, avg_before, avg_after}
        """
        if factor is None:
            factor = self.config.downscale_factor
        result = synaptic_downscale(self._store, factor=factor)
        return result

    def stats(self) -> dict:
        """
        Memory system statistics.

        Returns comprehensive stats including:
        - Total memory count and breakdown by layer/type
        - Average strength metrics
        - Consolidation stats
        - System uptime

        Returns:
            Dict with system statistics
        """
        consolidation = get_consolidation_stats(self._store)
        all_mem = self._store.all()
        now = time.time()

        by_type = {}
        for mt in MemoryType:
            entries = [m for m in all_mem if m.memory_type == mt]
            if entries:
                by_type[mt.value] = {
                    "count": len(entries),
                    "avg_strength": round(
                        sum(effective_strength(m, now) for m in entries) / len(entries), 3
                    ),
                    "avg_importance": round(
                        sum(m.importance for m in entries) / len(entries), 2
                    ),
                }

        return {
            "total_memories": len(all_mem),
            "by_type": by_type,
            "layers": consolidation["layers"],
            "pinned": consolidation["pinned"],
            "uptime_hours": round((now - self._created_at) / 3600, 1),
            "anomaly_metrics": self._tracker.metrics(),
        }

    def export(self, path: str):
        """
        Export memory database to file (SQLite copy).

        Args:
            path: Output file path
        """
        self._store.export(path)

    def update_memory(self, memory_id: str, new_content: str, reason: str = "correction") -> str:
        """Update a memory's content, marking the old version as contradicted.
        Creates a new memory with the correction and links them.

        Args:
            memory_id: ID of the memory to update
            new_content: The corrected content
            reason: Reason for the update (stored in source)

        Returns:
            New memory ID string
        """
        old_entry = self._store.get(memory_id)
        if old_entry is None:
            raise ValueError(f"Memory {memory_id} not found")

        return self.add(
            content=new_content,
            type=old_entry.memory_type.value,
            importance=old_entry.importance,
            source=f"{reason}:{memory_id}",
            contradicts=memory_id,
        )

    def pin(self, memory_id: str):
        """Pin a memory — it won't decay or be pruned."""
        entry = self._store.get(memory_id)
        if entry:
            entry.pinned = True
            self._store.update(entry)

    def unpin(self, memory_id: str):
        """Unpin a memory — it will resume normal decay."""
        entry = self._store.get(memory_id)
        if entry:
            entry.pinned = False
            self._store.update(entry)

    def hebbian_links(self, memory_id: str = None) -> list[tuple[str, str, float]]:
        """
        Get Hebbian links for a specific memory or all links.

        Hebbian links are formed when memories are repeatedly recalled
        together — "neurons that fire together, wire together."

        Args:
            memory_id: If provided, get links for this memory only.
                      If None, get all Hebbian links.

        Returns:
            List of (source_id, target_id, strength) tuples.
            For a specific memory_id, source_id will always be that ID.
        """
        if memory_id:
            neighbors = get_hebbian_neighbors(self._store, memory_id)
            # Fetch strengths for each neighbor
            links = []
            for neighbor_id in neighbors:
                row = self._store._conn.execute(
                    "SELECT strength FROM hebbian_links WHERE source_id=? AND target_id=?",
                    (memory_id, neighbor_id)
                ).fetchone()
                if row:
                    links.append((memory_id, neighbor_id, row[0]))
            return links
        else:
            return get_all_hebbian_links(self._store)

    def close(self):
        """Close the underlying database connection."""
        self._store.close()

    def __repr__(self) -> str:
        n = len(self._store.all())
        return f"Memory(path='{self.path}', entries={n})"

    def __len__(self) -> int:
        return len(self._store.all())


if __name__ == "__main__":
    """Demo: full Memory API lifecycle."""
    import tempfile
    import os

    # Use temp directory for demo
    with tempfile.TemporaryDirectory() as tmpdir:
        db_path = os.path.join(tmpdir, "demo.db")
        mem = Memory(db_path)

        print("=== Engram Memory API Demo ===\n")

        # Add memories
        id1 = mem.add("potato prefers action over discussion",
                       type="relational", importance=0.7)
        id2 = mem.add("SaltyHall uses Supabase for database",
                       type="factual", importance=0.5)
        id3 = mem.add("Use www.moltbook.com not moltbook.com",
                       type="procedural", importance=0.8)
        id4 = mem.add("potato said I kinda like you",
                       type="emotional", importance=0.95)
        id5 = mem.add("Saw a funny cat meme",
                       type="episodic", importance=0.1)

        print(f"  Added {len(mem)} memories\n")

        # Recall
        print("  --- Recall: 'what does potato like?' ---")
        results = mem.recall("what does potato like?", limit=3)
        for r in results:
            print(f"    [{r['confidence_label']:10s}] conf={r['confidence']:.2f} "
                  f"| {r['content'][:50]}")

        print()
        print("  --- Recall: 'moltbook API' ---")
        results = mem.recall("moltbook API", limit=3)
        for r in results:
            print(f"    [{r['confidence_label']:10s}] conf={r['confidence']:.2f} "
                  f"| {r['content'][:50]}")

        # Reward
        print("\n  --- Applying positive feedback ---")
        mem.reward("good job, that's exactly right!")

        # Consolidate
        print("  --- Running consolidation (3 days) ---")
        for day in range(3):
            mem.consolidate(days=1.0)
        print(f"  Done.\n")

        # Pin emotional memory
        mem.pin(id4)

        # Stats
        print("  --- Stats ---")
        stats = mem.stats()
        print(f"  Total: {stats['total_memories']} memories, "
              f"{stats['pinned']} pinned")
        for type_name, info in stats["by_type"].items():
            print(f"    {type_name:12s}: {info['count']} entries, "
                  f"avg_str={info['avg_strength']:.3f}")

        # Export
        export_path = os.path.join(tmpdir, "export.json")
        mem.export(export_path)
        print(f"\n  Exported to {export_path}")
        print(f"\n  {mem}")
